src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/FrozenDictionary.cs

```diff
@@ -105,14 +110,94 @@ public static class FrozenDictionary
                 // the Equals/GetHashCode methods to be devirtualized and possibly inlined.
                 if (ReferenceEquals(comparer, EqualityComparer<TKey>.Default))
                 {
-                    // In the specific case of Int32 keys, we can optimize further to reduce memory consumption by using
-                    // the underlying FrozenHashtable's Int32 index as the keys themselves, avoiding the need to store the
-                    // same keys yet again.
-                    return typeof(TKey) == typeof(int) ?
-                        (FrozenDictionary<TKey, TValue>)(object)new Int32FrozenDictionary<TValue>((Dictionary<int, TValue>)(object)source) :
-                        new ValueTypeDefaultComparerFrozenDictionary<TKey, TValue>(source);
+#if NET7_0_OR_GREATER
+                    static FrozenDictionary<TKey, TValue> PickIntegerDictionary<TInt>(Dictionary<TKey, TValue> source)
+                        where TInt : struct, IBinaryInteger<TInt>
+                    {
+                        var k = (TInt[])(object)source.Keys.ToArray();
+                        var v = source.Values.ToArray();
+
+                        Array.Sort(k, v);
+
+                        var min = k[0];
+                        var max = k[k.Length - 1];
+                        var range = ulong.CreateTruncating(max - min);
+
+                        if (range == (ulong)k.Length - 1)
+                        {
+                            return (FrozenDictionary<TKey, TValue>)(object)new ClosedRangeIntegerFrozenDictionary<TInt, TValue>(k, v);
+                        }
+                        else if (k.Length <= MaxItemsInSmallIntegerDictionary)
+                        {
+                            return (FrozenDictionary<TKey, TValue>)(object)new SmallIntegerFrozenDictionary<TInt, TValue>(k, v);
```

Comment:
Is this really specific to integers?  Couldn't it apply more generally to any small collection of `IComparable<T>`s?

Reply:
There are the two general-purpose SmallFrozenDictionary and SmallFrozenSet types which deal with arbitrary types. That seems sufficient to me.